File: internal/core/adapter.go
package core

import "time"

type SandboxSubmission struct {
	TransactionID     string  `json:"transaction_id"`
	ProviderReference string  `json:"provider_reference"`
	RouteID           RouteID `json:"route_id"`
}

type SandboxOutcome string

const (
	SandboxAccepted SandboxOutcome = "accepted"
	SandboxRejected SandboxOutcome = "rejected"
	SandboxTimeout  SandboxOutcome = "timeout"
	SandboxFailed   SandboxOutcome = "failed"
)

type SandboxSubmissionResult struct {
	Outcome            SandboxOutcome    `json:"outcome"`
	SimulatedLatencyMS uint64            `json:"simulated_latency_ms"`
	ProviderReference  string            `json:"provider_reference"`
	Callbacks          []SandboxCallback `json:"callbacks"`
}

type SandboxCallback struct {
	TransactionID     string           `json:"transaction_id"`
	ProviderReference string           `json:"provider_reference"`
	State             TransactionState `json:"state"`
	OccurredAt        time.Time        `json:"occurred_at"`
	Duplicate         bool             `json:"duplicate"`
}

func (c SandboxCallback) AsEvent(eventID string) TransactionEvent {
	message := "sandbox provider callback"
	if c.Duplicate {
		message = "duplicate sandbox provider callback"
	}

	return NewTransactionEvent(eventID, c.TransactionID, c.State, EventSourceProvider, message)
}

type SandboxProviderAdapter struct {
	ProviderID         ProviderID `json:"provider_id"`
	LatencyMS          uint64     `json:"latency_ms"`
	FailureRateBps     uint16     `json:"failure_rate_bps"`
	RejectionRateBps   uint16     `json:"rejection_rate_bps"`
	TimeoutRateBps     uint16     `json:"timeout_rate_bps"`
	DuplicateCallbacks uint8      `json:"duplicate_callbacks"`
}

func NewAlwaysAcceptSandboxAdapter(providerID ProviderID) SandboxProviderAdapter {
	return SandboxProviderAdapter{ProviderID: providerID}
}

func (a SandboxProviderAdapter) Submit(submission SandboxSubmission) SandboxSubmissionResult {
	bucket := deterministicBucket(submission.TransactionID)
	timeoutCutoff := min(a.TimeoutRateBps, 10_000)
	rejectionCutoff := timeoutCutoff + min(a.RejectionRateBps, 10_000-timeoutCutoff)
	failureCutoff := rejectionCutoff + min(a.FailureRateBps, 10_000-rejectionCutoff)

	outcome := SandboxAccepted
	switch {
	case bucket < timeoutCutoff:
		outcome = SandboxTimeout
	case bucket < rejectionCutoff:
		outcome = SandboxRejected
	case bucket < failureCutoff:
		outcome = SandboxFailed
	}

	return SandboxSubmissionResult{
		Outcome:            outcome,
		SimulatedLatencyMS: a.LatencyMS,
		ProviderReference:  submission.ProviderReference,
		Callbacks:          callbacksFor(submission, outcome, a.DuplicateCallbacks),
	}
}

func callbacksFor(submission SandboxSubmission, outcome SandboxOutcome, duplicateCallbacks uint8) []SandboxCallback {
	state := StateProviderAccepted
	if outcome != SandboxAccepted {
		state = StateFailed
	}

	callback := SandboxCallback{
		TransactionID:     submission.TransactionID,
		ProviderReference: submission.ProviderReference,
		State:             state,
		OccurredAt:        time.Now().UTC(),
	}

	callbacks := []SandboxCallback{callback}
	for range duplicateCallbacks {
		duplicate := callback
		duplicate.Duplicate = true
		callbacks = append(callbacks, duplicate)
	}

	return callbacks
}

func deterministicBucket(value string) uint16 {
	hash := uint32(0)
	for _, b := range []byte(value) {
		hash = hash*31 + uint32(b)
	}
	return uint16(hash % 10_000)
}

File: crates/imsi-core/src/adapter.rs
use crate::lifecycle::{EventSource, TransactionEvent, TransactionState};
use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};

#[derive(Clone, Debug, Deserialize, Eq, PartialEq, Serialize)]
pub struct SandboxSubmission {
    pub transaction_id: String,
    pub provider_reference: String,
    pub route_id: String,
}

#[derive(Clone, Copy, Debug, Deserialize, Eq, PartialEq, Serialize)]
#[serde(rename_all = "snake_case")]
pub enum SandboxOutcome {
    Accepted,
    Rejected,
    Timeout,
    Failed,
}

#[derive(Clone, Debug, Deserialize, Eq, PartialEq, Serialize)]
pub struct SandboxSubmissionResult {
    pub outcome: SandboxOutcome,
    pub simulated_latency_ms: u64,
    pub provider_reference: String,
    pub callbacks: Vec<SandboxCallback>,
}

#[derive(Clone, Debug, Deserialize, Eq, PartialEq, Serialize)]
pub struct SandboxCallback {
    pub transaction_id: String,
    pub provider_reference: String,
    pub state: TransactionState,
    pub occurred_at: DateTime<Utc>,
    pub duplicate: bool,
}

impl SandboxCallback {
    pub fn as_event(&self, event_id: impl Into<String>) -> TransactionEvent {
        TransactionEvent::new(
            event_id,
            self.transaction_id.clone(),
            self.state,
            EventSource::Provider,
            if self.duplicate {
                "duplicate sandbox provider callback"
            } else {
                "sandbox provider callback"
            },
        )
    }
}

#[derive(Clone, Debug, Deserialize, Eq, PartialEq, Serialize)]
pub struct SandboxProviderAdapter {
    pub provider_id: String,
    pub latency_ms: u64,
    pub failure_rate_bps: u16,
    pub rejection_rate_bps: u16,
    pub timeout_rate_bps: u16,
    pub duplicate_callbacks: u8,
}

impl SandboxProviderAdapter {
    pub fn always_accept(provider_id: impl Into<String>) -> Self {
        Self {
            provider_id: provider_id.into(),
            latency_ms: 0,
            failure_rate_bps: 0,
            rejection_rate_bps: 0,
            timeout_rate_bps: 0,
            duplicate_callbacks: 0,
        }
    }

    pub fn submit(&self, submission: SandboxSubmission) -> SandboxSubmissionResult {
        let bucket = deterministic_bucket(&submission.transaction_id);
        let timeout_cutoff = self.timeout_rate_bps.min(10_000);
        let rejection_cutoff =
            timeout_cutoff + self.rejection_rate_bps.min(10_000 - timeout_cutoff);
        let failure_cutoff =
            rejection_cutoff + self.failure_rate_bps.min(10_000 - rejection_cutoff);

        let outcome = if bucket < timeout_cutoff {
            SandboxOutcome::Timeout
        } else if bucket < rejection_cutoff {
            SandboxOutcome::Rejected
        } else if bucket < failure_cutoff {
            SandboxOutcome::Failed
        } else {
            SandboxOutcome::Accepted
        };

        SandboxSubmissionResult {
            outcome,
            simulated_latency_ms: self.latency_ms,
            provider_reference: submission.provider_reference.clone(),
            callbacks: callbacks_for(&submission, outcome, self.duplicate_callbacks),
        }
    }
}

fn callbacks_for(
    submission: &SandboxSubmission,
    outcome: SandboxOutcome,
    duplicate_callbacks: u8,
) -> Vec<SandboxCallback> {
    let state = match outcome {
        SandboxOutcome::Accepted => TransactionState::ProviderAccepted,
        SandboxOutcome::Rejected | SandboxOutcome::Failed | SandboxOutcome::Timeout => {
            TransactionState::Failed
        }
    };

    let callback = SandboxCallback {
        transaction_id: submission.transaction_id.clone(),
        provider_reference: submission.provider_reference.clone(),
        state,
        occurred_at: Utc::now(),
        duplicate: false,
    };

    let mut callbacks = vec![callback.clone()];
    for _ in 0..duplicate_callbacks {
        let mut duplicate = callback.clone();
        duplicate.duplicate = true;
        callbacks.push(duplicate);
    }

    callbacks
}

fn deterministic_bucket(value: &str) -> u16 {
    let hash = value.bytes().fold(0u32, |acc, byte| {
        acc.wrapping_mul(31).wrapping_add(u32::from(byte))
    });
    (hash % 10_000) as u16
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn sandbox_adapter_can_simulate_acceptance_and_duplicate_callbacks() {
        let adapter = SandboxProviderAdapter {
            provider_id: "sandbox".into(),
            latency_ms: 250,
            failure_rate_bps: 0,
            rejection_rate_bps: 0,
            timeout_rate_bps: 0,
            duplicate_callbacks: 2,
        };

        let result = adapter.submit(SandboxSubmission {
            transaction_id: "txn-1".into(),
            provider_reference: "sandbox-ref-1".into(),
            route_id: "route-1".into(),
        });

        assert_eq!(result.outcome, SandboxOutcome::Accepted);
        assert_eq!(result.simulated_latency_ms, 250);
        assert_eq!(result.callbacks.len(), 3);
        assert_eq!(
            result.callbacks[0].state,
            TransactionState::ProviderAccepted
        );
        assert!(result.callbacks[1].duplicate);
    }

    #[test]
    fn sandbox_adapter_can_force_timeout() {
        let adapter = SandboxProviderAdapter {
            provider_id: "sandbox".into(),
            latency_ms: 10_000,
            failure_rate_bps: 0,
            rejection_rate_bps: 0,
            timeout_rate_bps: 10_000,
            duplicate_callbacks: 0,
        };

        let result = adapter.submit(SandboxSubmission {
            transaction_id: "txn-timeout".into(),
            provider_reference: "sandbox-ref-2".into(),
            route_id: "route-1".into(),
        });

        assert_eq!(result.outcome, SandboxOutcome::Timeout);
        assert_eq!(result.callbacks[0].state, TransactionState::Failed);
    }
}
